eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -29,39 +29,51 @@ import scodec.bits.ByteVector
+  case class ChannelRelayPacket(add: UpdateAddHtlc, payload: PaymentOnion.ChannelRelayPayload, nextPacket: OnionRoutingPacket) extends RelayPacket {
     val relayFeeMsat: MilliSatoshi = add.amountMsat - payload.amountToForward
     val expiryDelta: CltvExpiryDelta = add.cltvExpiry - payload.outgoingCltv
   }
   /** We must relay the payment to a remote node. */
-  case class NodeRelayPacket(add: UpdateAddHtlc, outerPayload: Onion.FinalPayload, innerPayload: Onion.NodeRelayPayload, nextPacket: OnionRoutingPacket) extends RelayPacket
+  case class NodeRelayPacket(add: UpdateAddHtlc, outerPayload: PaymentOnion.FinalPayload, innerPayload: PaymentOnion.NodeRelayPayload, nextPacket: OnionRoutingPacket) extends RelayPacket
   // @formatter:on
 
-  case class DecodedOnionPacket[T <: Onion.PacketType](payload: T, next: OnionRoutingPacket)
+  case class DecodedOnionPacket[T <: PaymentOnion.PacketType](payload: T, next: OnionRoutingPacket)
 
-  private[payment] def decryptOnion[T <: Onion.PacketType : ClassTag](paymentHash: ByteVector32, privateKey: PrivateKey)(packet: OnionRoutingPacket, packetType: Sphinx.OnionRoutingPacket[T])(implicit log: LoggingAdapter): Either[FailureMessage, DecodedOnionPacket[T]] =
-    packetType.peel(privateKey, paymentHash, packet) match {
+  private[payment] def decryptPaymentOnion(paymentHash: ByteVector32, privateKey: PrivateKey, packet: OnionRoutingPacket)(implicit log: LoggingAdapter): Either[FailureMessage, DecodedOnionPacket[PaymentOnion.PaymentPacket]] =
+    Sphinx.peel(privateKey, paymentHash, packet) match {
       case Right(p@Sphinx.DecryptedPacket(payload, nextPacket, _)) =>
-        (OnionCodecs.perHopPayloadCodecByPacketType(packetType, p.isLastPacket).decode(payload.bits): @unchecked) match {
-          case Attempt.Successful(DecodeResult(perHopPayload: T, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
-          case Attempt.Failure(e: OnionCodecs.MissingRequiredTlv) => Left(e.failureMessage)
+        PaymentOnionCodecs.paymentOnionPerHopPayloadCodec(p.isLastPacket).decode(payload.bits) match {
+          case Attempt.Successful(DecodeResult(perHopPayload: PaymentOnion.PaymentPacket, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
+          case Attempt.Failure(e: OnionRoutingCodecs.MissingRequiredTlv) => Left(e.failureMessage)
+          // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
+          // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
+          case Attempt.Failure(_) => Left(InvalidOnionPayload(UInt64(0), 0))
+        }
+      case Left(badOnion) => Left(badOnion)
+    }
+
+  private[payment] def decryptTrampolineOnion(paymentHash: ByteVector32, privateKey: PrivateKey, packet: OnionRoutingPacket)(implicit log: LoggingAdapter): Either[FailureMessage, DecodedOnionPacket[PaymentOnion.TrampolinePacket]] =
```

Comment:
I've chosen to duplicate this code to be able to get rid of the `unchecked` pattern matching, I think having the compiler verify that it's exhaustive is better than avoiding a few lines of duplication.

I was hoping I could get the best of both worlds, but haven't been able to. If you find a refactoring that achieves it, I'm interested!

Reply:
```scala
private[payment] def decryptOnion[T <: PaymentOnion.PacketType](perHopPayloadCodec : (Boolean) => Codec[T])(paymentHash: ByteVector32, privateKey: PrivateKey, packet: OnionRoutingPacket)(implicit log: LoggingAdapter): Either[FailureMessage, DecodedOnionPacket[T]] =
    Sphinx.peel(privateKey, paymentHash, packet) match {
      case Right(p@Sphinx.DecryptedPacket(payload, nextPacket, _)) =>
        perHopPayloadCodec(p.isLastPacket).decode(payload.bits) match {
          case Attempt.Successful(DecodeResult(perHopPayload, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
          case Attempt.Failure(e: OnionRoutingCodecs.MissingRequiredTlv) => Left(e.failureMessage)
          // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
          // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
          case Attempt.Failure(_) => Left(InvalidOnionPayload(UInt64(0), 0))
        }
      case Left(badOnion) => Left(badOnion)
    }
```
And then you use `decryptOnion(PaymentOnionCodecs.paymentOnionPerHopPayloadCodec)` and `decryptOnion(PaymentOnionCodecs.trampolineOnionPerHopPayloadCodec)`.

Reply:
Good catch! I had done almost exactly the same thing, with a tiny difference, I was doing:

```
case Attempt.Successful(DecodeResult(perHopPayload: T, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
```

But that wasn't working because of type erasure (I had to add the `unchecked`).
I didn't even try removing the `: T` type constraint because I thought it was this method's responsibility to enforce the type bounds, but with your proposal it's actually the `perHopPayloadCodec` function that enforces it and it's all fine.

Thanks!

Reply:
Done in https://github.com/ACINQ/eclair/pull/2052/commits/17ed855e8320d51e78793dc80824ed38f771097d